tools/cargos/src/main.rs

```diff
@@ -0,0 +1,253 @@
+// Cargo.toml section.
+//
+// A common task that is especially annoying to perform by hand is to update
+// trybuild's stderr files. Using this script:
+//
+//   TRYBUILD=overwrite ./cargos +all test --workspace
+
+use std::{
+    env::{args, vars},
+    fmt,
+    io::{stdin, Read as _},
+    process::{exit, Command},
+};
+
+use serde_json::{Map, Value};
+
+#[derive(Debug)]
+enum Error {
+    NoArguments,
+    UnrecognizedArgument(String),
+    MissingToolchainVersion,
+    UnrecognizedToolchain(String),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NoArguments => write!(f, "No arguments provided"),
+            Self::UnrecognizedArgument(arg) => write!(f, "Unrecognized argument: '{arg}'"),
+            Self::MissingToolchainVersion => write!(f, "No toolchain version specified after '--version'"),
+            Self::UnrecognizedToolchain(name) => write!(f, "Unrecognized toolchain name: `{name}` (options are 'msrv', 'stable', and 'nightly')")
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+struct Versions {
+    msrv: String,
+    stable: String,
+    nightly: String,
+    build_rs: Map<String, Value>,
+}
+
+impl Versions {
+    fn get(&self, name: &str) -> Result<&str, Error> {
+        Ok(match name {
+            "msrv" => &self.msrv,
+            "stable" => &self.stable,
+            "nightly" => &self.nightly,
+            _ => self
+                .build_rs
+                .get(name)
+                .ok_or(Error::UnrecognizedToolchain(name.to_string()))
+                .map(|value| value.as_str().unwrap())?,
+        })
+    }
+}
+
+fn get_toolchain_versions() -> Versions {
+    // NOTE(#547): We set `CARGO_TARGET_DIR` here because `cargo metadata`
+    // sometimes causes the `cargo-metadata` crate to be rebuilt from source
+    // using the default toolchain. This has the effect of clobbering any
+    // existing build artifacts from whatever toolchain the user has specified
+    // (e.g., `+nightly`), causing the subsequent `cargo` invocation to rebuild
+    // unnecessarily. By specifying a separate build directory here, we ensure
+    // that this never clobbers the build artifacts used by the later `cargo`
+    // invocation.
+    //
+    // In CI, make sure to use the default stable toolchain. If we're testing on
+    // our MSRV, then we also have our MSRV toolchain installed. As of this
+    // writing, our MSRV is low enough that the correspoding Rust toolchain's
+    // Cargo doesn't know about the `rust-version` field, and so if we were to
+    // use Cargo with that toolchain, `pkg-meta` would return `null` when asked
+    // to retrieve the `rust-version` field. This also requires `RUSTFLAGS=''`
+    // to override any unstable `RUSTFLAGS` set by the caller.
+    let output = Command::new("rustup")
+        .args(["run", "stable", "cargo", "metadata", "--format-version", "1"])
```

Comment:
Why not `cargo +stable` like we did before? Is there an advantage to calling via `rustup`?

Reply:
I think there's some shell trickery going on here. When I run `cargo +stable metadata` in powershell, it acts as expected. However, if I `Command::new("cargo").args(["+stable", "metadata"])` in Rust, I get:

```
stdout: 
stderr: error: no such command: `+stable`

        Cargo does not handle `+toolchain` directives.
        Did you mean to invoke `cargo` through `rustup` instead?
```

I suspect there's some kind of aliasing going on here where "cargo" as a command resolves to a specific toolchain's `cargo.exe` binary, but from the shell it resolves to rustup's "cargo" dispatcher binary.

Reply:
Gotcha, that makes sense.